### DISCUSSION TITLE
Python code formatting with black

### DISCUSSION THREAD

**Discussion by perara-chen** [zed-industries/zed#6597] [Feature Requests]
It is just only been a few days since I started using Zed and I love it. I could setup the Zed for my rust learning out of the box with very few configurations. But I am having problem with formatting Python code on save. I have already read the documentation and already installed black. But I still can't use it. I would like to request the help to setup Python for Zed from seniors. Thanks.

My settings.json:
```
// Zed settings
//
// For information on how to configure Zed, see the Zed
// documentation: https://zed.dev/docs/configuring-zed
//
// To see all of Zed's default settings without changing your
// custom settings, run the `open default settings` command
// from the command palette or from `Zed` application menu.
{
  // Appearence settings
  "terminal": {
    "dock": "bottom"
  },
  "theme": "Andromeda",
  // Text editor settings
  "buffer_font_size": 15,
  "buffer_font_family": "Zed Mono",
  "format_on_save": "on",
  // Vim mode
  "vim_mode": true,
  // Text wrapping settings
  "wrap_guides": [
    80,
    120
  ],
  "show_wrap_guides": true,
  // Autosave settings
  "autosave": "on_focus_change",
  // LSP settings
  // Rust
  "lsp": {
    "rust-analyzer": {
      "initialization_options": {
        "rustfmt": {
          "extraArgs": [
            "--config",
            "wrap_comments=true"
          ]
        }
      }
    }
  },
  // Code formatting
  // Python
  "language_overrides": {
    "Python": {
      "formatter": {
        "external": {
          "command": "black",
          "arguments": [
            "-t",
            "py311",
            "--line-length=80"
          ]
        }
      }
    }
  }
}
```

**Comment by levlaz** [comment]
Could you share more about what happens when you save a file? Does it do anything, does it make the formatting incorrectly, does it throw any errors?

**Reply by perara-chen** [reply]
It does not throw any errors or something else. There is no effects or errors. So, I don't know what is going on. I also tried with full path for black. But it did not solve the problem. By the way, I installed black with MacPorts.

**Comment by dbuxton** [comment]
BTW I had this issue and the solution is simple - you need to just pass `"-"` as the first argument, because that makes `black` read from stdin. Hope that helps someone. 

e.g.

```
"language_overrides": {
  "Python": {
    "formatter": { "external": { "command": "black", "arguments": ["-"] } }
  }
}
```

**Comment by guilhermecaldas** [comment]
Here is my working settings, except by the `-t py311` that you can add later:
```JSON
"languages": {
    "Python": {
      "format_on_save": {
        "external": {
          "command": "/opt/homebrew/bin/black",
          "arguments": ["-", "--line-length", "80"]
        }
      }
    }
  }
  ```

Note: At least for me, Zed had to be restarted after every setting update in order to apply changes.

**Comment by einstein95** [comment]
```json
  "languages": {
    "Python": {
      "format_on_save": "on",
      "formatter": {
        "external": {
          "command": "black",
          "arguments": ["--stdin-filename", "{buffer_path}", "-"]
        }
      }
    }
  }
```
Throwing in my 2c that `--stdin-filename` is good to include as from documentation:
```
  --stdin-filename TEXT           The name of the file when passing it through
                                  stdin. Useful to make sure Black will
                                  respect the --force-exclude option on some
                                  editors that rely on using stdin.
```

**Comment by breathe** [comment]
How do we get zed to find the black installed into the project's venv dir?

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/community-links
